src/vision/status_quo/barbara_plays_with_async.md

```diff
@@ -0,0 +1,387 @@
+
+After watching the development of rust async/await (by following
+discussions on [/r/rust][reddit] and the [internals] forums), she wants
+to start to play around with writing async code.  Before starting on any real
+project, she starts with a "playground" where she can try to write some simple
+async rust code to see how it feels and how it compares to how async code feels
+in other languages she knows (like C# and JavaScript).
+
+She starts by opening a blank project in VSCode with [rust-analyzer].  Because she's
+been following the overall state of rust async, she knows that she needs a runtime,
+and quickly decides to use tokio, because she knows its quite popular and well documented.
+
+After looking the long length of the [tokio tutorial], she decides to not read
+most of it right now, and tries to dive right in to writing code.  But she does
+look at the "[Hello Tokio]" section that shows what feature flags are required by tokio:
+
+```toml
+[dependencies]
+tokio = { version = "1", features = ["full"] }
+```
+
+Poking around the tokio API docs in search of something to play with, she sees
+a simple future that looks interesting: the [`sleep`][tokio sleep] future
+that will wait for a certain duration to elapse before resolving.
+
+Borrowing again from the "Hello Tokio" tutorial to make sure she has the correct
+spelling for the tokio macros, she writes up the following code:
+
+```rust
+#[tokio::main]
+pub async fn main() {
+    let mut rng = thread_rng();
+    let t = Uniform::new(100, 5000);
+
+    let mut futures = Vec::new();
+    for _ in 0..10 {
+        let delay = rng.sample(t);
+        futures.push(tokio::time::sleep(Duration::from_millis(delay)));
+    }
+    println!("Created 10 futures");
+
+    for f in futures {
+        f.await;
+    }
+
+    println!("Done waiting for all futures");
+}
+```
+
+This very first version she wrote compiled on the first try and had no errors
+when running it.  Barbara was pleased about this.
+
+However, this example is pretty boring.  The program just sits there for a few
+seconds doing nothing, and giving no hints about what it's actually doing.
+So for the next iteration, Barbara wants to have a message printed out
+when each future is resolved.  She tries this code at first:
+
+```rust
+let mut futures = Vec::new();
+for _ in 0..10 {
+    let delay = rng.sample(t);
+    futures.push(tokio::time::sleep(Duration::from_millis(delay)).then(|_| {
+        println!("Done!");
+    }));
+}
+println!("Created 10 futures");
+```
+
+But the compiler gives this error:
+
+```
+error[E0277]: `()` is not a future
+  --> src\main.rs:13:71
+   |
+13 |         futures.push(tokio::time::sleep(Duration::from_millis(delay)).then(|_| {
+   |                                                                       ^^^^ `()` is not a future
+   |
+   = help: the trait `futures::Future` is not implemented for `()`
+```
+
+Even though the error is pointing at the `then` function, Barbara pretty quickly
```

Comment:
Hard to say how it would be if I were editing the code, but fwiw I stared at this quite a long time before reading on to understand what was happening. I could see this bug being quite confusing. =)

Reply:
This is a great point.  For me personally, I was able to draw upon experiences from years ago trying to use futures in a pre-async world, where you can get some pretty gnarly type errors.   But I don't know if we want to give Barbara those same experiences, or if we want to have her struggle a bit more with this particular error.

Also, when reviewing this code again I realized that the `then` method is only available as part of the `futures` crate, however, we haven't yet introduced this crate as this point in the story.  I kinda feel like this part of the story should be rewritten to fix this inconsistency.  What do you think?

Reply:
Yeah, it seems ok to have her know about futures already and go on the lookout for combinators.

Reply:
👍 ok gonna mark this as 'resolved' for now